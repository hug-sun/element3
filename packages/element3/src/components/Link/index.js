import ElLink from './src/Link.vue'

/* istanbul ignore next */
ElLink.install = function (app) {
  app.component(ElLink.name, ElLink)
}

export { ElLink }
