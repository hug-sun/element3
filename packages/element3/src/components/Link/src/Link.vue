<template>
  <a
    :class="classes"
    :href="disabled ? null : href"
    v-bind="$attrs"
    @click="handleClick"
  >
    <i v-if="icon" :class="icon"></i>
    <span v-if="$slots.default" class="el-link--inner">
      <slot></slot>
    </span>
  </a>
</template>

<script>
export default {
  name: 'ElLink',
  props: {
    type: {
      type: String,
      default: 'default'
    },
    disabled: {
      type: Boolean,
      default: false
    },
    underline: {
      type: Boolean,
      default: true
    },
    href: String,
    icon: String
  },
  emits: ['click'],
  setup(props, { emit }) {
    const classes = useClasses(props)

    const handleClick = (event) => {
      if (props.disabled) return
      if (props.href) return

      emit('click', event)
    }
    return {
      classes,
      handleClick
    }
  }
}
const useClasses = (props) => {
  return [
    'el-link',
    props.type ? `el-link--${props.type}` : '',
    props.disabled && 'is-disabled',
    props.underline && !props.disabled && 'is-underline'
  ]
}
</script>

<style></style>
