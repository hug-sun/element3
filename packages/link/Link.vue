<template>
  <a
    :class="[
      'el-link',
      type ? `el-link--${type}` : '',
      disabled && 'is-disabled',
      underline && !disabled && 'is-underline'
    ]"
    :href="disabled ? null : href"
    v-bind="$attrs"
    @click="handleClick"
  >
    <i :class="icon" v-if="icon"></i>

    <span v-if="$slots.default" class="el-link--inner">
      <slot></slot>
    </span>

    <slot v-if="$slots.icon" name="icon"></slot>
  </a>
</template>

<script>
export default {
  name: 'ElLink',

  props: {
    type: {
      type: String,
      default: 'default'
    },
    underline: {
      type: Boolean,
      default: true
    },
    disabled: Boolean,
    href: String,
    icon: String
  },
  emits: ['click'],
  setup(props, { emit }) {
    const handleClick = (evt) => {
      if (props.disabled) return
      if (props.href) return

      emit('click', evt)
    }

    return {
      handleClick
    }
  }
}
</script>
