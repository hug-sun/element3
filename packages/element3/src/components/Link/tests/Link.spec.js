import { mount } from '@vue/test-utils'
import Link from '../src/Link.vue'
describe('Link', () => {
  it('snapshot', () => {
    const wrapper = mount(Link)
    expect(wrapper.element).toMatchSnapshot()
  })
  describe('props', () => {
    it('initialize the Link component', () => {
      const wrapper = mount(Link)
      expect(wrapper.find('i').exists()).toBe(false)
      expect(wrapper.find('span').exists()).toBe(false)
    })
    it('should show content', () => {
      const content = 'Link'
      const wrapper = mount(Link, {
        slots: {
          default: content
        }
      })
      expect(wrapper.text()).toContain('Link')
    })

    it('set the type, link displays the corresponding style', () => {
      const type = 'primary'
      const wrapper = mount(Link, {
        props: {
          type
        }
      })
      expect(wrapper.classes()).toContain(`el-link--${type}`)
    })

    it('set the disabled, link displays the corresponding style', () => {
      const disabled = true
      const wrapper = mount(Link, {
        props: {
          disabled
        }
      })
      expect(wrapper.classes()).toContain('is-disabled')
    })

    it('set the underline, link displays the corresponding style', () => {
      const underline = true
      const wrapper = mount(Link, {
        props: {
          underline
        }
      })

      expect(wrapper.classes()).toContain('is-underline')
    })

    it('set the href', () => {
      const href = 'https://element3-ui.com/'
      const wrapper = mount(Link, {
        props: {
          href
        }
      })
      expect(wrapper.attributes('href')).toBe(href)
    })

    it('set the icon, link displays the corresponding style', () => {
      const icon = 'el-icon-search'
      const wrapper = mount(Link, {
        props: {
          icon
        }
      })
      const i = wrapper.find('i')
      expect(i.exists()).toBe(true)
      expect(i.classes()).toContain(icon)
    })

    it('should get target attr value', () => {
      const wrapper = mount(Link, {
        props: {
          target: '_blank'
        }
      })
      expect(wrapper.attributes('target')).toBe('_blank')
    })
  })

  describe('click', () => {
    it('should captured click events emitted via click', () => {
      const wrapper = mount(Link)
      wrapper.trigger('click')
      expect(wrapper.emitted('click')).toBeTruthy()
    })

    describe("can't captured click event emitted", () => {
      //TODO vue 3.0.7 的时候 vtu 对收集 emit 有问题，这个测试等 vtu 解决后打开
      it.skip('when disabled prop is equal to true', () => {
        const wrapper = mount(Link, {
          props: {
            disabled: true
          }
        })
        wrapper.trigger('click')
        expect(wrapper.emitted('click')).toBeFalsy()
      })

      //TODO vue 3.0.7 的时候 vtu 对收集 emit 有问题，这个测试等 vtu 解决后打开
      it.skip('when href prop is to be truthy', () => {
        const wrapper = mount(Link, {
          props: {
            href: 'https://element3-ui.com/'
          }
        })
        wrapper.trigger('click')
        expect(wrapper.emitted('click')).toBeFalsy()
      })
    })
  })
})
