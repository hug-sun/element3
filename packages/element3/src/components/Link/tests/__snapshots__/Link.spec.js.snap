// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link snapshot 1`] = `
<a
  class="el-link el-link--default is-underline"
>
  <!--v-if-->
  <!--v-if-->
</a>
`;
